neo/Ledger/Blockchain.cs

```diff
@@ -0,0 +1,705 @@
+        }
+
+        private void OnNewHeaders(Header[] headers)
+        {
+            using (Snapshot snapshot = GetSnapshot())
+            {
+                foreach (Header header in headers)
+                {
+                    if (header.Index - 1 >= header_index.Count) break;
+                    if (header.Index < header_index.Count) continue;
+                    if (!header.Verify(snapshot)) break;
+                    header_index.Add(header.Hash);
+                    snapshot.Blocks.Add(header.Hash, new BlockState
+                    {
+                        SystemFeeAmount = 0,
+                        TrimmedBlock = header.Trim()
+                    });
+                    snapshot.HeaderHashIndex.GetAndChange().Hash = header.Hash;
+                    snapshot.HeaderHashIndex.GetAndChange().Index = header.Index;
+                }
+                SaveHeaderHashList(snapshot);
+                snapshot.Commit();
+            }
+            UpdateCurrentSnapshot();
+            system.TaskManager.Tell(new TaskManager.HeaderTaskCompleted(), Sender);
+        }
+
+        private RelayResultReason OnNewTransaction(Transaction transaction)
+        {
+            const int MemoryPoolSize = 50000;
+            if (transaction.Type == TransactionType.MinerTransaction)
+                return RelayResultReason.Invalid;
+            if (ContainsTransaction(transaction.Hash))
+                return RelayResultReason.AlreadyExists;
+            if (!transaction.Verify(currentSnapshot, mem_pool.Values))
+                return RelayResultReason.Invalid;
+            mem_pool.TryAdd(transaction.Hash, transaction);
+            if (mem_pool.Count > MemoryPoolSize)
+            {
+                UInt256[] delete = mem_pool.Values.AsParallel()
+                    .OrderBy(p => p.NetworkFee / p.Size)
+                    .ThenBy(p => p.NetworkFee)
+                    .ThenBy(p => new BigInteger(p.Hash.ToArray()))
+                    .Take(mem_pool.Count - MemoryPoolSize)
+                    .Select(p => p.Hash)
+                    .ToArray();
+                foreach (UInt256 hash in delete)
+                    mem_pool.TryRemove(hash, out _);
+            }
+            if (!mem_pool.ContainsKey(transaction.Hash))
+                return RelayResultReason.OutOfMemory;
+            system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = transaction });
+            return RelayResultReason.Succeed;
+        }
+
+        private void OnPersistCompleted(Block block)
+        {
+            block_cache.Remove(block.Hash);
+            foreach (Transaction tx in block.Transactions)
+                mem_pool.TryRemove(tx.Hash, out _);
+            Transaction[] remain = mem_pool.Values.ToArray();
+            mem_pool.Clear();
+            foreach (Transaction tx in remain)
+            {
+                if (!tx.Verify(currentSnapshot, mem_pool.Values)) continue;
+                mem_pool.TryAdd(tx.Hash, tx);
```

Comment:
I updated the code, and put unverified transactions in the memory pool back into the message queue after the block is persisted. So it will relay them again after reverifying.